src/choice_map.hpp

```diff
@@ -158,154 +58,49 @@ class ChoiceMap {
-    SizeVector node_ids;
-    bool root_check = false;
-    BoolVector leaf_check(dag_.TaxonCount(), false);
-    // Node map for checking node connectivity: Array is ordered as [left_child,
-    // right_child, parent].
-    using NodeMap = std::map<size_t, std::array<bool, 3>>;
-    NodeMap nodemap_check;
+  bool TreeMaskIsValid(const TreeMask &tree_mask, const bool is_quiet = true) const;
 
-    for (size_t i = 0; i < tree_mask.size(); i++) {
-      const auto edge_idx = tree_mask[i];
-      const auto &edge = dag_.GetDAGEdge(edge_idx);
-      const auto &parent_node = dag_.GetDAGNode(edge.GetParent());
-      const auto &child_node = dag_.GetDAGNode(edge.GetChild());
-      // Check if edge goes to root.
-      if (dag_.IsNodeRoot(parent_node.Id())) {
-        if (root_check == true) {
-          return false;
-        }
-        root_check = true;
-      }
-      // Check if edge goes to leaf.
-      if (dag_.IsNodeLeaf(child_node.Id())) {
-        const auto taxon_id = child_node.Id();
-        if (leaf_check.at(taxon_id) == true) {
-        }
-        leaf_check.at(taxon_id) = true;
-      }
-      // Update node map. If a node already has parent or child, it is an invalid tree.
-      for (const auto node_id : {parent_node.Id(), child_node.Id()}) {
-        if (nodemap_check.find(node_id) == nodemap_check.end()) {
-          nodemap_check.insert({node_id, {false, false, false}});
-        }
-      }
-      const size_t which_child =
-          (edge.GetSubsplitClade() == SubsplitClade::Left) ? 0 : 1;
-      if (nodemap_check[parent_node.Id()][which_child] == true) {
-        return false;
-      }
-      nodemap_check[parent_node.Id()][which_child] = true;
-      if (nodemap_check[child_node.Id()][2] == true) {
-        return false;
-      }
-      nodemap_check[child_node.Id()][2] = true;
-    }
-    // Check if all nodes are fully connected.
-    for (const auto &[node_id, connections] : nodemap_check) {
-      // Check children.
-      if (!(connections[0] || connections[1])) {
-        // If one child is not connected, neither should be.
-        if (connections[0] ^ connections[1]) {
-          return false;
-        }
-        if (!dag_.IsNodeLeaf(node_id)) {
-          return false;
-        }
-      }
-      // Check parent.
-      if (!connections[2]) {
-        if (!dag_.IsNodeRoot(node_id)) {
-          return false;
-        }
-      }
-    }
-    // Check if spans root and all leaf nodes.
-    if (!root_check) {
-      return false;
-    }
-    for (size_t i = 0; i < leaf_check.size(); i++) {
-      if (!leaf_check[i]) {
-        return false;
-      }
-    }
+  // Output TreeMask to string.
+  std::string TreeMaskToString(const TreeMask &tree_mask) const;
+  std::string ExpandedTreeMaskToString(const ExpandedTreeMask &tree_mask) const;
 
-    return true;
-  }
+  // ** Tree
+
+  // Extract Tree from DAG based on edge choices to find best tree with given
+  // central edge.
+  // - Makes two passes:
```

Comment:
These are implementation docs, which I love, but they belong in the cpp file.

Reply:
Moved.